pkg/route/ingress/ingress.go

```diff
@@ -717,6 +675,63 @@ func generateRouteName(base string) string {
 	return fmt.Sprintf("%s%s", base, utilrand.String(randomLength))
 }
 
+func tlsConfigForIngress(
+	ingress *networkingv1beta1.Ingress,
+	rule *networkingv1beta1.IngressRule,
+	potentiallyNilTLSSecret *corev1.Secret,
+) *routev1.TLSConfig {
+	if !tlsEnabled(ingress, rule) {
+		return nil
+	}
+	// Edge: May have cert
+	// Re-Encrypt: May have cert
+	// Passthrough: Must not have cert
+	terminationPolicy := terminationPolicyForIngress(ingress)
+	tlsConfig := &routev1.TLSConfig{
+		Termination:                   terminationPolicy,
+		InsecureEdgeTerminationPolicy: routev1.InsecureEdgeTerminationPolicyRedirect,
+	}
+	if terminationPolicy != routev1.TLSTerminationPassthrough && potentiallyNilTLSSecret != nil {
+		tlsConfig.Certificate = string(potentiallyNilTLSSecret.Data[corev1.TLSCertKey])
+		tlsConfig.Key = string(potentiallyNilTLSSecret.Data[corev1.TLSPrivateKeyKey])
+	}
+	return tlsConfig
+}
+
+func tlsEnabled(ingress *networkingv1beta1.Ingress, rule *networkingv1beta1.IngressRule) bool {
+	if val := ingress.Annotations[terminationPolicyAnnotationKey]; val == string(routev1.TLSTerminationPassthrough) || val == string(routev1.TLSTerminationReencrypt) || val == string(routev1.TLSTerminationEdge) {
+		return true
+	}
+	if len(ingress.Spec.TLS) == 1 && reflect.DeepEqual(ingress.Spec.TLS[0], networkingv1beta1.IngressTLS{}) {
+		return true
+	}
+	_, referencesSecret := referencesSecret(ingress, rule.Host)
+	return referencesSecret
```

Comment:
No, because an Ingress can have multiple `IngressRule`s which each lead to a route. In all other cases, we match the `IngressRule` to a `IngressTLS` via its `Host` attribute, the empty one is a special magic way of enabling it without cert for all Ingresses.

Reply:
Hm, all right.  Is it sufficient to check only the first item in `spec.tls`, or could it make sense for an ingress to specify one rule that has a nonempty TLS config and another rule that uses a second, empty TLS config?

Reply:
The empty tls doesn't allow any kind of matching to rules, this is why using it is only allowed if its the only element and it then applies to all of them.

Reply:
What about the following ingress?
```yaml
spec:
  tls:
  - secretName: tls
    hosts:
    - foo.com
  - {}
  rules:
  - host: foo.com
    http:
      paths:
      - backend:
          serviceName: service-secure
          servicePort: 8443
  - host: foo.apps.mycluster.com
    http:
      paths:
      - backend:
          serviceName: service-secure
          servicePort: 8443
```
I am able to create the ingress, so it seems at least to be valid.  I would expect the above ingress to use the "tls" secret for foo.com and the default certificate for foo.apps.mycluster.com.

Reply:
Updated it to accept any empty tlsconfig regardless of what other tlsconfigs are present